File: 2015/10/rust/src/main.rs
use std::fs;


fn solve(data: &str, n_iter: usize) -> usize {
    let mut text = data.to_string();

    for _ in 0..n_iter {
        let mut out_str = String::new();
        let mut chars = text.chars().peekable();
        let mut count = 1;

        let mut prev = chars.next().unwrap();

        while let Some(&next) = chars.peek() {
            if next == prev {
                count += 1;
            } else {
                out_str.push_str(&count.to_string());
                out_str.push(prev);
                count = 1;
            }
            prev = chars.next().unwrap();
        }
        out_str.push_str(&count.to_string());
        out_str.push(prev);
        text = out_str;
    }
    text.len()
}

fn main() {
    let data = fs::read_to_string("../input.txt").expect("Failed to read").trim().to_string();
    let sep = "=".repeat(20);
    println!("{} Part 1 {}", sep, sep);
    let p1 = solve(&data, 40);
    println!("{}",p1);
    println!("{} Part 2 {}", sep, sep);
    let p2 = solve(&data, 50);
    println!("{}",p2);
}

File: 2015/10/go/main.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

func solve(data string, nIter int) int {
	text := data

	for i := 0; i < nIter; i++ {
		var outStr strings.Builder
		chars := []rune(text)
		count := 1
		prev := chars[0]

		for j := 1; j < len(chars); j++ {
			if chars[j] == prev {
				count++
			} else {
				outStr.WriteString(strconv.Itoa(count))
				outStr.WriteRune(prev)
				count = 1
			}
			prev = chars[j]
		}
		outStr.WriteString(strconv.Itoa(count))
		outStr.WriteRune(prev)
		text = outStr.String()
	}

	return len(text)
}

func main() {
	data, err := os.ReadFile("../input.txt")
	if err != nil {
		panic("Failed to read file")
	}
	input := strings.TrimSpace(string(data))

	sep := strings.Repeat("=", 20)
	fmt.Println(sep, "Part 1", sep)
	p1 := solve(input, 40)
	fmt.Println(p1)

	fmt.Println(sep, "Part 2", sep)
	p2 := solve(input, 50)
	fmt.Println(p2)
}
